backend/src/v2/driver/driver.go

```diff
@@ -749,6 +756,33 @@ func DAG(ctx context.Context, opts Options, mlmd *metadata.Client) (execution *E
 	ecfg.ParentDagID = dag.Execution.GetID()
 	ecfg.IterationIndex = iterationIndex
 	ecfg.NotTriggered = !execution.WillTrigger()
+
+	// Handle writing output parameters to MLMD.
+	outputParameters := opts.Component.GetDag().GetOutputs().GetParameters()
+	glog.V(4).Info("outputParameters: ", outputParameters)
+	for _, value := range outputParameters {
+		outputParameterKey := value.GetValueFromParameter().OutputParameterKey
+		producerSubTask := value.GetValueFromParameter().ProducerSubtask
+		glog.V(4).Info("outputParameterKey: ", outputParameterKey)
+		glog.V(4).Info("producerSubtask: ", producerSubTask)
+
+		outputParameterMap := map[string]interface{}{
+			"output_parameter_key": outputParameterKey,
+			"producer_subtask":     producerSubTask,
+		}
+
+		outputParameterStruct, _ := structpb.NewValue(outputParameterMap)
```

Comment:
how come you didn't use a [DagOutputParameterSpec](https://github.com/kubeflow/pipelines/blob/83dcf1a60919f5bcc0c644c8fdff94ad686cad07/api/v2alpha1/pipeline_spec.proto#L157) like you did for [DagOutputArtifactSpec](https://github.com/kubeflow/pipelines/pull/11196/files#diff-87ec6fcace7656e2c64443ea8b5386b8d54215eab715ab193fc0e2c1bfe939d1R139)?

Reply:
The logic we added to the mlmd client handles converting the **artifact** struct into a format suitable for storage in the database. Our reasoning was that that was an implementation detail that wasn't really relevant to the driver. That logic holds true for the output parameters as well, but we never refactored it to apply the principle. We'll move that logic out of `driver.go` and into `client.go`. Good call!